File: backend/app/schemas/notes.py
from __future__ import annotations
from pydantic import BaseModel


class NoteUpsertRequest(BaseModel):
    year: int
    month: int
    service_name: str
    note: str
    note_date: str   # YYYY-MM-DD
    resource_id: str | None = None
    resource_name: str | None = None
    filter_name: str | None = None
    cloud: str = "aws"
    cloud_account: str = "default"
    # Legacy field — accepted for backward compat
    aws_profile: str | None = None


class NoteItem(BaseModel):
    id: int
    year: int
    month: int
    service_name: str
    note: str
    note_date: str
    resource_id: str | None
    resource_name: str | None
    filter_name: str | None
    cloud: str
    cloud_account: str
    aws_profile: str
    created_at: str

    model_config = {"from_attributes": True}


class NotesListResponse(BaseModel):
    data: list[NoteItem]
    count: int

File: frontend/src/api/notes.ts
import client from './client'
import type { NoteItem } from '../types/billing'

export async function fetchNotes(
  year: number,
  month: number,
  cloud = 'aws',
  cloud_account = 'default',
): Promise<NoteItem[]> {
  const { data } = await client.get(`/notes/${year}/${month}`, {
    params: { cloud, cloud_account },
  })
  return data.data
}

export async function fetchNotesByRange(
  fromDate: string,
  toDate: string,
  cloud = 'aws',
  cloud_account = 'default',
): Promise<NoteItem[]> {
  const { data } = await client.get('/notes', {
    params: { from_date: fromDate, to_date: toDate, cloud, cloud_account },
  })
  return data.data
}

export async function upsertNote(payload: {
  year: number
  month: number
  service_name: string
  note: string
  note_date: string
  resource_id?: string | null
  resource_name?: string | null
  filter_name?: string | null
  cloud?: string
  cloud_account?: string
}): Promise<NoteItem> {
  const { data } = await client.post('/notes', payload)
  return data
}

export async function deleteNote(id: number): Promise<void> {
  await client.delete(`/notes/${id}`)
}
